hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -28,6 +28,65 @@ def _filter_missing_method(self, filter_missing: bool, name: str, ret_type: Hail
             collection = self.filter(hl.is_defined)
         return collection._method(name, ret_type, *args)
 
+    def __getattr__(self, item):
+        return CollectionExpression.__getitem__(self, item)
+
+    @typecheck_method(item=oneof(str))
+    def __getitem__(self, item):
+        """Get a field from each struct in this collection.
+
+        Examples
+        --------
+
+        Note, in the examples below ``hl.eval`` is used to convert a hail
+        expression to a python value so that it prints nicely. It is not
+        necssary to use ``hl.eval`` inside ``annotate_rows`` or similar
+        methods.
+
+        >>> x = [hl.struct(a='foo', b=3), hl.struct(a='bar', b=4)]
+        >>> hl.eval(x.a)
+        ['foo', 'bar']
+
+        >>> a = [hl.struct(b=[hl.struct(inner=1),
+        ...                   hl.struct(inner=2)]),
+        ...      hl.struct(b=[hl.struct(inner=3)])]
+        >>> hl.eval(a.b)
+        [[hl.struct(inner=1), hl.struct(inner=2)], [hl.struct(inner=3)]]
+        >>> hl.eval(a.b.inner)
+        [[1, 2], [3]]
+        >>> hl.eval(hl.flatten(a.b).inner)
+        [1, 2, 3]
+        >>> hl.eval(hl.flatten(a.b.inner))
+        [1, 2, 3]
+
+        Parameters
+        ----------
+        item : :obj:`str`
+            Field name
+
+        Returns
+        -------
+        :class:`.CollectionExpression`
+            A collection formed by getting the given field for each struct in
+            this collection
+        """
+
+        etype = self.dtype.element_type
+        if isinstance(etype, hl.tstruct):
+            return hl.struct(
+                **{k: self.map(lambda x: x[k]) for k in etype}).__getitem__(item)
+        try:
+            if isinstance(etype, (hl.tarray, hl.tset)):
+                return self.map(lambda x: x.__getitem__(item))
+        except TypeError as err:
+            if "Cannot use 'collection.foo' or" not in str(err):
+                raise
+        raise TypeError(
+            f"Cannot use 'collection.foo' or 'collection['foo']' on "
```

Comment:
Do you actually want `foo` in here? Could you use `item`, or does the recursion mess with this?

Reply:
oh sure yeah item

Reply:
done